### ISSUE TITLE
Should.ThrowAsync<T> does seem to work correctly if the underlying Func has a CancellationToken which cancelled

### ISSUE THREAD

**Issue by PureKrome** [shouldly/shouldly#457]
Hi 👋 

so I'm not sure if this is "by design" and I've not set up my stuff correctly ... or it's a bug.

So I have a simple service I'm testing out. This service accepts a `CancellationToken`. If the service fails to complete in 5 seconds, the CancellationToken cancels, which ends up throwing an exception.

The current code in our unit tests _works_ 👍 

```
var wasCancelled = false;
try
{
    // This hardcoded class is made to throw a `TaskCanceledException` after 5 seconds.
   // The implimentation of this method is a simple `Task.Delay` for 15 seconds.
    await clientService.GetJsonAsync("doesnt-matter");
}
catch (TaskCanceledException)
{
    wasCancelled = true;
}

// Assert.
wasCancelled.ShouldBeTrue();

```

Now, I tried to refactor this into `Shouldly` code like this:

- `var result = await Should.ThrowAsync<TaskCancelledException>(clientService.GetJsonAsync("doesnt-matter"));`
- `var result = await Should.ThrowAsync<TaskCancelledException>(() => clientService.GetJsonAsync("doesnt-matter"));`

and both fail. No exception is thrown by Shouldly.

e.g.

```
Shouldly.ShouldAssertException
Task `clientService.GetJsonAsync("doesnt-matter")`
    should throw
System.Threading.Tasks.TaskCanceledException
    but did not

... <snip >

System.Threading.Tasks.TaskCanceledException
Task is cancelled
  Exception doesn't have a stacktrace
```

---

So I thought I might try and add a test method to the `Shouldly.Tests` projects to see if I can help repo this:

```
[Fact]
public async Task ShouldThrowAWobbly_WhenATaskIsCancelled()
{
    // Arrange.
    // Cancel this calling code after 5 seconds.
    var cancellationTokenSource = new CancellationTokenSource(5);
    var task = Task.Delay(TimeSpan.FromSeconds(10), cancellationTokenSource.Token);

    // Act.
    var result = await Should.ThrowAsync<Exception>(() => task);

    // Assert.
    result.ShouldNotBeNull();
}
```

☝️ So the above code is trying to capture _any_ `Exception` when a Task is cancelled.

Here's a screen shot of the error:

![image](https://user-images.githubusercontent.com/899878/29392978-8c8d176e-8344-11e7-9bdd-98deffea8d37.png)

Interesting:

- ✅ It's figured out that the task is `Cancelled`.  
- ❌ But it now checks to see if an exception was thrown _by my code_ which it has not ... and as such, sorta assplodes.

So I'm not sure if this is something I've not setup up correctly with Shouldly or it's a legit bug.

**Comment by paulomorgado** [comment]
This fails:

```C#
await Should.ThrowAsync<>(async () => await test());
```

But this doesn't;

```C#
try
{
    await test();
}
catch (OperationCanceledException)
{
    return;
}

throw new Exception("Test failed!");
```

**Comment by JosephWoodward** [comment]
This has now been fixed as part of #560 and will be in the next beta release.
